libats-slick/src/test/scala/com/advancedtelematic/libats/slick/db/SlickExtensionsSpec.scala

```diff
@@ -118,10 +131,39 @@ class SlickExtensionsSpec extends FunSuite with Matchers with ScalaFutures with
     f.failed.futureValue shouldBe Error
   }
 
+  test("handleForeignKeyError throws the expected error if one FK is missing") {
+    val b = Book(50, "Beyond Good and Evil", Some("81781857261933"))
+    val bm = BookMeta(b.id, 51, 0)
+    val f = db.run {
+      (books += b).andThen(bookMeta += bm).handleForeignKeyError(Error, {
+        case NoReferencedRow(msg) if msg.contains("book_id_fk") => ErrorBookIdFk
```

Comment:
But why a partial function? Why not model `ErrorBookIdFk` and `ErrorBookTagFk` as case classes with an `unapply` method?

Reply:
But then if the matching over the errors is done in libats, it needs to know about the errors from all the services?

Reply:
I do not see an advantage of moving it to libats.
```
db.run(...).handleForeignKeyError(Error, {
        case NoReferencedRow(msg) if msg.contains("book_id_fk") => ErrorBookIdFk
        case NoReferencedRow(msg) if msg.contains("book_tag_fk") => ErrorBookTagFk
      }
```
vs.
```
db.run(...).recover {
  case Failure(ErrorBookIdFk) => ???
  case Failure(ErrorBookTagFk) => ???
}

Reply:
But we wouldn't have `case Failure(ErrorBookIdFk) => ???`, we would have `case Failure(SQLIntegrityConstraintViolationException) => ???`. I'm assuming `ErrorBookIdFk` is defined in a service and libats doesn't know about it.

Reply:
`ErrorBookIdFk` will not be defined it `libats`. 
You can have `case Failure(ErrorBookIdFk) =>` in service.
The examples from https://github.com/advancedtelematic/libats/pull/91#discussion_r255034967 are examples how handling will look like in a service. And my point is there is no advantage in adding the `.handleForeignKeyError` to the `libats`

Reply:
Mmh but `ErrorBookIdFk` is what we want to return eventually, not what we can match in the `recover` method. If I'm trying to do it like this in a service, say campaigner, I'm having to do something like:
```
(Schema.campaigns += campaign).recover {
        case Failure(e: SQLIntegrityConstraintViolationException) if e.getErrorCode == 1452 =>
          if (e.getMessage.contains("fk_1"))
            DBIO.failed(FKError1)
          else if (e.getMessage.contains("fk_2"))
            DBIO.failed(FKError2)
          else DBIO.failed(e)
      }
```

Reply:
You can have the `unapply` for `ErrorBookTagFk` in the service, and make it unapply from a `SQLIntegrityConstraintViolationException` which is what @koshelev is saying I guess.

I like `        case NoReferencedRow(msg) if msg.contains("book_id_fk") => ErrorBookIdFk` because the id of the constraint being violated is right in the method that will need to handle it, rather than somewhere else where `unapply` for `ErrorBookIdFk` is defined.

But I can live with the `unapply` method, the only thing I don't want is to pass a `Map` of stuff `libats` will try to make sense of. `PF+constraint id` or `unapply+constraint id` for me is not that important.